### PR TITLE
wan federation via mesh gateways

### PR DESCRIPTION
**Pull request by rboyer**
Dependent upon:
- https://github.com/hashicorp/memberlist/pull/210
- https://github.com/hashicorp/serf/pull/590

Fixes: https://github.com/hashicorp/consul/issues/6356

### Summary

I've sometimes joked and referred to this as a Möbius strip of code due to the fact that low-level components (serf/memberlist) are connected to high-level components (the catalog and mesh-gateways) in a twisty maze of references which make it hard to dive into. With that in mind here's a high level summary of what you'll find in the PR:

There are several distinct chunks of code that are affected:

* new flags and config options for the server

* retry join WAN is slightly different

* retry join code is shared to discover primary mesh gateways from secondary datacenters

* because retry join logic runs in the *agent* and the results of that
  operation for primary mesh gateways are needed in the *server* there are
  some methods like `RefreshPrimaryGatewayFallbackAddresses` that must occur
  at multiple layers of abstraction just to pass the data down to the right
  layer.

* new cache type `FederationStateListMeshGatewaysName` for use in `proxycfg/xds` layers

* the function signature for RPC dialing picked up a new required field (the
  node name of the destination)

* several new RPCs for manipulating a FederationState object:
  `FederationState:{Apply,Get,List,ListMeshGateways}`

* 3 read-only internal APIs for debugging use to invoke those RPCs from curl

* raft and fsm changes to persist these FederationStates

* replication for FederationStates as they are canonically stored in the
  Primary and replicated to the Secondaries.

* a special derivative of anti-entropy that runs in secondaries to snapshot
  their local mesh gateway `CheckServiceNodes` and sync them into their upstream
  FederationState in the primary (this works in conjunction with the
  replication to distribute addresses for all mesh gateways in all DCs to all
  other DCs)

* a "gateway locator" convenience object to make use of this data to choose
  the addresses of gateways to use for any given RPC or gossip operation to a
  remote DC. This gets data from the "retry join" logic in the agent and also
  directly calls into the FSM.

* RPC (`:8300`) on the server sniffs the first byte of a new connection to
  determine if it's actually doing native TLS. If so it checks the ALPN header
  for protocol determination (just like how the existing system uses the
  type-byte marker).

* 2 new kinds of protocols are exclusively decoded via this native TLS
  mechanism: one for ferrying "packet" operations (udp-like) from the gossip
  layer and one for "stream" operations (tcp-like). The packet operations
  re-use sockets (using length-prefixing) to cut down on TLS re-negotiation
  overhead.

* the server instances specially wrap the `memberlist.NetTransport` when running
  with gateway federation enabled (in a `wanfed.Transport`). The general gist is
  that if it tries to dial a node in the SAME datacenter (deduced by looking
  at the suffix of the node name) there is no change. If dialing a DIFFERENT
  datacenter it is wrapped up in a TLS+ALPN blob and sent through some mesh
  gateways to eventually end up in a server's :8300 port.

* a new flag when launching a mesh gateway via `consul connect envoy` to
  indicate that the servers are to be exposed. This sets a special service
  meta when registering the gateway into the catalog.

* `proxycfg/xds` notice this metadata blob to activate additional watches for
  the FederationState objects as well as the location of all of the consul
  servers in that datacenter.

* `xds:` if the extra metadata is in place additional clusters are defined in a
  DC to bulk sink all traffic to another DC's gateways. For the current
  datacenter we listen on a wildcard name (`server.<dc>.consul`) that load
  balances all servers as well as one mini-cluster per node
  (`<node>.server.<dc>.consul`)

* the `consul tls cert create` command got a new flag (`-node`) to help create
  an additional SAN in certs that can be used with this flavor of federation.

### TODO

- [ ] remove remaining random TODOs in the PR
- [x] envoy integration test
- [x] default primary_datacenter to the current datacenter during config parsing in the agent, not just in the server
- [x] fedstate AE in the primary should not do RPC, as it doesn't have a replication token
- [x] consider adding another flag to `consul tls cert create` to add the node-name SAN
- [ ] verify that server certs containing `*.server.DC.DOMAIN` work for wanfed
- [ ] write envoy integ test asserting that cross-DC connect upstreams work via gateways AFTER enabling wanfed via gateways
- [x] issue for better handling non-graceful rotation of mesh gateways in the primary DC
- [x] Documentation will happen elsewhere: https://github.com/hashicorp/consul/issues/7416

### CODE REVIEW

File: agent/consul/leader_routine_manager.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ func (m *LeaderRoutineManager) StartWithContext(parentCtx context.Context, name
 	go func() {
 		err := routine(ctx)
 		if err != nil && err != context.DeadlineExceeded && err != context.Canceled {
-			m.logger.Printf("[ERROR] leader: %s routine exited with error: %v", name, err)
+			m.logger.Printf("[ERR] leader: %s routine exited with error: %v", name, err)
```
Comment (rboyer) [inline agent/consul/leader_routine_manager.go]: This is an unrelated fix.

File: agent/pool/conn.go [hunk 2]
```diff
@@ -13,7 +36,7 @@ const (
 	RPCGossip              = 6
 	// RPCTLSInsecure is used to flag RPC calls that require verify
 	// incoming to be disabled, even when it is turned on in the
-	// configuration. At the time of writing there is only AutoEncryt.Sign
+	// configuration. At the time of writing there is only AutoEncrypt.Sign
```
Comment (rboyer) [inline agent/pool/conn.go]: This is an unrelated fix.

File: agent/config/runtime.go [hunk 3]
```diff
@@ -983,14 +1002,14 @@ type RuntimeConfig struct {
 	// attempts on agent start. The minimum allowed value is 1 second and
 	// the default is 30s.
 	//
-	// hcl: retry_join = "duration"
+	// hcl: retry_interval = "duration"
```
Comment (rboyer) [inline agent/config/runtime.go]: Unrelated typo fix.

File: agent/config/runtime.go [hunk 4]
```diff
@@ -983,14 +1002,14 @@ type RuntimeConfig struct {
...
 	RetryJoinIntervalLAN time.Duration
 
 	// RetryJoinIntervalWAN specifies the amount of time to wait in between join
 	// attempts on agent start. The minimum allowed value is 1 second and
 	// the default is 30s.
 	//
-	// hcl: retry_join_wan = "duration"
+	// hcl: retry_interval_wan = "duration"
```
Comment (rboyer) [inline agent/config/runtime.go]: Unrelated typo fix.

File: agent/consul/rpc.go [hunk 5]
```diff
@@ -128,6 +148,63 @@ func (s *Server) handleConn(conn net.Conn, isTLS bool) {
...
+
+	case pool.ALPN_RPCMultiplexV2:
+		s.handleMultiplexV2(tlsConn)
+
+	case pool.ALPN_RPCSnapshot:
+		s.handleSnapshotConn(tlsConn)
+
+	case pool.ALPN_WANGossipPacket:
+		// No need to defer the conn.Close() here, the Ingest methods do that.
+		if err := transport.IngestPacket(tlsConn, tlsConn.RemoteAddr(), time.Now()); err != nil {
```
Comment (rboyer) [inline agent/consul/rpc.go]: maybe length prefix this and don't close

File: agent/consul/leader_datacenter_config_ae.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	)
+	err := s.blockingQuery(
+		queryOpts,
+		&queryMeta,
+		func(ws memdb.WatchSet, state *state.Store) error {
+			// Get the existing stored version of this config that has replicated down.
+			// We could phone home to get this but that would incur extra WAN traffic
+			// when we already have enough information locally to figure it out
+			// (assuming that our replicator is still functioning).
+			idx1, prev, err := state.DatacenterConfigGet(ws, s.config.Datacenter)
```
Comment (rboyer) [inline agent/consul/leader_datacenter_config_ae.go]: Just block over the wan?

Comment (rboyer) [reply agent/consul/leader_datacenter_config_ae.go]: Or block with a timer again?

Comment (rboyer) [reply agent/consul/leader_datacenter_config_ae.go]: Version number? PrimaryModifyIndex?

File: agent/http_register.go [hunk 7]
```diff
@@ -78,6 +78,9 @@ func init() {
 	registerEndpoint("/v1/coordinate/nodes", []string{"GET"}, (*HTTPServer).CoordinateNodes)
 	registerEndpoint("/v1/coordinate/node/", []string{"GET"}, (*HTTPServer).CoordinateNode)
 	registerEndpoint("/v1/coordinate/update", []string{"PUT"}, (*HTTPServer).CoordinateUpdate)
+	registerEndpoint("/v1/datacenter-configs", []string{"GET"}, (*HTTPServer).DatacenterConfigList)
```
Comment (rboyer) [inline agent/http_register.go]: Move reading to internal.

File: agent/http_register.go [hunk 8]
```diff
@@ -78,6 +78,9 @@ func init() {
...
+	registerEndpoint("/v1/datacenter-config/", []string{"GET", "PUT", "DELETE"}, (*HTTPServer).DatacenterConfigCRUD)
```
Comment (rboyer) [inline agent/http_register.go]: No writes.

File: agent/http_register.go (hunk 7 above)
Comment (rboyer) [inline agent/http_register.go]: - `s/DatacenterConfig/MeshGatewayMap/`
- `s/DatacenterConfig/FederationState/` <--

File: agent/config/config.go [hunk 9]
```diff
@@ -571,9 +573,10 @@ type AutoEncrypt struct {
 type Connect struct {
 	// Enabled opts the agent into connect. It should be set on all clients and
 	// servers in a cluster for correct connect operation.
-	Enabled    *bool                  `json:"enabled,omitempty" hcl:"enabled" mapstructure:"enabled"`
-	CAProvider *string                `json:"ca_provider,omitempty" hcl:"ca_provider" mapstructure:"ca_provider"`
-	CAConfig   map[string]interface{} `json:"ca_config,omitempty" hcl:"ca_config" mapstructure:"ca_config"`
+	Enabled                         *bool                  `json:"enabled,omitempty" hcl:"enabled" mapstructure:"enabled"`
+	CAProvider                      *string                `json:"ca_provider,omitempty" hcl:"ca_provider" mapstructure:"ca_provider"`
+	CAConfig                        map[string]interface{} `json:"ca_config,omitempty" hcl:"ca_config" mapstructure:"ca_config"`
+	MeshGatewayWANFederationEnabled *bool                  `json:"enable_mesh_gateway_wan_federation" hcl:"enable_mesh_gateway_wan_federation" mapstructure:"enable_mesh_gateway_wan_federation"`
```
Comment (rboyer) [inline agent/config/config.go]: Q: is this name too verbose?

Comment (s-christoff) [reply agent/config/config.go]: Maybe just - `WANFedEnabled` or `WANFederationEnabled` - do we need the Mesh Gateway part?

Comment (rboyer) [reply agent/config/config.go]: Well technically normal multidc consul _is_ federation via the wan.

Comment (banks) [reply agent/config/config.go]: It is a bit verbose but I can't think of a shorter version that actually communicates the effect of setting it correctly. We could shorten `Federation` to `Fed` but in general in HashiCorp products our convention is not to abbreviate name or config options. So I guess leave it like this!

Comment (lkysow) [reply agent/config/config.go]: Do we need WAN? What other type of federation is there?

File: agent/consul/federation_state_endpoint.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+	if args.State == nil || args.State.Datacenter == "" {
+		return fmt.Errorf("invalid request: missing federation state datacenter")
+	}
+
+	switch args.Op {
+	case structs.FederationStateUpsert:
+		if !c.srv.router.HasDatacenter(args.State.Datacenter) &&
+			!c.srv.config.DisableFederationStateDatacenterNameValidation {
+			return fmt.Errorf("cannot upsert federation state for datacenter that does not exist: %s", args.State.Datacenter)
```
Comment (banks) [inline agent/consul/federation_state_endpoint.go]: Is it worth commenting why this is important to enforce? Given that you need to have operator:write already, you could do a lot worse than fill up the state store with arbitrary Fed States for non-existent DCs. I can imagine there could be ordering issues where this would at least cause errors for a while until things resolve (but might be wrong about that).

Just wondering if there are other strong reasons to enforce this I'm missing?

File: agent/consul/federation_state_replication.go [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+type FederationStateReplicator struct {
+	srv *Server
+}
+
+func (r *FederationStateReplicator) SingularNoun() string { return "federation state" }
+func (r *FederationStateReplicator) PluralNoun() string   { return "federation states" }
+func (r *FederationStateReplicator) MetricName() string   { return "federation-state" }
+
+func (r *FederationStateReplicator) FetchRemote(lastRemoteIndex uint64) (int, interface{}, uint64, error) {
```
Comment (banks) [inline agent/consul/federation_state_replication.go]: nit: I think these methods are implementing the replicator interface right? I'm partial to adding Doc comments like `// FetchRemote implements IndexReplicatorDelegate `

File: agent/consul/gateway_locator.go [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+	"github.com/hashicorp/consul/ipaddr"
+	"github.com/hashicorp/consul/lib"
+	"github.com/hashicorp/go-hclog"
+	memdb "github.com/hashicorp/go-memdb"
+)
+
+// GatewayLocator assists in selecting an appropriate mesh gateway when wan
+// federation via mesh gateways is enabled.
+//
+// NOTE: this is ONLY for RPC and gossip so we only need to track local/primary
```
Comment (banks) [inline agent/consul/gateway_locator.go]: It's not super clear to me why we'd only need primary and local for gossip - traditionally gossip needs a full mesh of connectivity.

File: agent/consul/gateway_locator.go [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+
+	if primary && len(addrs) == 0 {
+		addrs = g.PrimaryGatewayFallbackAddresses()
+	}
+
+	return addrs
+}
+
+// RefreshPrimaryGatewayFallbackAddresses is used to update the list of current
+// fallback addresses for locating mesh gateways in the primary datacenter.
```
Comment (banks) [inline agent/consul/gateway_locator.go]: Maybe worth noting the semantics of the int returned. It's not super obvious to me why the caller would care if the update was effectively a no-op or not or why they'd need to know the current number of addrs if it was  changed considering that they passed that array in the first place. I.e. why isn't it just a bool?

File: agent/consul/gateway_locator.go [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+		return ""
+	case 1:
+		return items[0]
+	default:
+		idx := int(rand.Int31n(int32(len(items))))
+		return items[idx]
+	}
+}
+
+type serverDelegate interface {
```
Comment (banks) [inline agent/consul/gateway_locator.go]: 🤔 why decouple this component from the server code but then keep it in the same package? I don't mind particularly but we often use this pattern to lift a component into a different package without circular dependencies/simpler testing.

I'm guessing by the `s := g // TODO` aliases that this was once built into the server directly but now isn't for a good reason though - perhaps that's worth documenting here?

File: agent/consul/gateway_locator.go [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+			"primary", primaryAddrs,
+			"local", localAddrs,
+		)
+	}
+
+	if primaryReady {
+		select {
+		case _, open := <-g.primaryGatewaysReadyCh:
+			if open {
+				close(g.primaryGatewaysReadyCh)
```
Comment (banks) [inline agent/consul/gateway_locator.go]: Does this ever fire? If the chan is open given that nothing is ever sent to it, wont the recv block and so the default path always be taken? That makes it look like this close will never happen.

But perhaps the `_, open :=` variant changes the behavior of recv somehow that I'm missing? This is more Go trivia for my info because from tests it looks like this must work!

File: agent/consul/gateway_locator.go [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+}
+
+func retainGateways(full structs.CheckServiceNodes) structs.CheckServiceNodes {
+	out := make([]structs.CheckServiceNode, 0, len(full))
+	for _, csn := range full {
+		if csn.Service.Meta[structs.MetaWANFederationKey] != "1" {
+			continue
+		}
+
+		// only keep healthy ones
```
Comment (banks) [inline agent/consul/gateway_locator.go]: 🤔 do you think we need a "panic threshold" where if there are multiple gateways and ~all~ more than X% are unhealthy we panic and just route to them anyway on the logic that it's better to try than fail completely?

File: agent/router/serf_flooder.go [hunk 17]
```diff
@@ -93,17 +93,19 @@ func FloodJoins(logger hclog.Logger, addrFn FloodAddrFn, portFn FloodPortFn,
 			}
 		}
 
+		globalServerName := fmt.Sprintf("%s.%s", server.Name, server.Datacenter)
+
 		// Do the join!
-		n, err := globalSerf.Join([]string{addr}, true)
+		n, err := globalSerf.Join([]string{globalServerName + "/" + addr}, true)
```
Comment (banks) [inline agent/router/serf_flooder.go]: Is this code only called in the Mesh gateway path now? What happens here if memberlist doesn't have its new node names mode enabled which I think is the case when not using mesh gateways right?

File: agent/structs/federation_state.go [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+)
+
+type FederationStateOp string
+
+const (
+	FederationStateUpsert FederationStateOp = "upsert"
+	FederationStateDelete FederationStateOp = "delete"
+)
+
+type FederationStateRequest struct {
```
Comment (banks) [inline agent/structs/federation_state.go]: Doc comments for context on these structs would be nice for later readers - we're pretty OK at doing that throughout structs package generally.

File: agent/structs/structs.go [hunk 19]
```diff
@@ -98,6 +99,10 @@ const (
 	// MetaSegmentKey is the node metadata key used to store the node's network segment
 	MetaSegmentKey = "consul-network-segment"
 
+	// MetaWANFederationKey is the mesh gateway metadata key that indicates a
+	// mesh gateway is usable for wan federation.
+	MetaWANFederationKey = "wan-federation"
```
Comment (banks) [inline agent/structs/structs.go]: Every other reserved meta key I think has the prefix `consul-` since that is our documented reserved prefix. I suggest we prefix this too just for consistency.

File: agent/xds/endpoints.go [hunk 20]
```diff
@@ -308,6 +377,8 @@ func makeLoadAssignmentEndpointGroup(
 		}, true
 	}
 
+	// TODO(rb): shouldn't we omit gateways that are unhealthy?
```
Comment (banks) [inline agent/xds/endpoints.go]: I'd say no - if we just tell Envoy they are unhealthy it can make more sensible choices like enforce it's panic threshold if more than 50% are healthy or similar. Saves us from building that kind of logic in here...

File: agent/consul/rpc.go [hunk 21]
```diff
@@ -257,6 +370,63 @@ func (s *Server) handleSnapshotConn(conn net.Conn) {
...
+		select {
+		case <-s.shutdownCh:
+			return nil
+		default:
+		}
+
+		// Note: if we need to change this format to have additional header
+		// information we can just negotiate a different ALPN protocol instead
+		// of needing any sort of version field here.
+		prefixLen, err := readUint32(conn, wanfed.GossipPacketMaxIdleTime)
```
Comment (banks) [inline agent/consul/rpc.go]: Oh one thing we should def do here for DOS prevention is enforce a sane max packet size.

If this len ends up being bigger than say 4MB (this is gRPC's default max RPC size that I proposed elsewhere adding to Consul RPC and plan to soon) we should just error the connection. I'm _pretty_ sure that's way bigger than Serf would ever need but we should probably check.

Even if we want to be super conservative and set it to 100MB or something I think a limit is better than no limit!

File: agent/consul/gateway_locator.go (hunk 12 above)
Comment (rboyer) [reply agent/consul/gateway_locator.go]: This is only keeping track of the gateways that the consul servers will directly need to dial.

- During secondary bootstrap you need to directly dial the primary DC's gateways.
- After secondary bootstrap all requests to through the local DC's gateway which does the job of routing the request to the correct gateway in a remote DC.

File: agent/router/serf_flooder.go (hunk 17 above)
Comment (rboyer) [reply agent/router/serf_flooder.go]: Understanding join addresses that are name-slash-prefixed is an update to the library that is independent of the toggle flag, so because the library is updated this is fine with or without fed enabled.

File: agent/structs/structs.go (hunk 19 above)
Comment (rboyer) [reply agent/structs/structs.go]: I wanted to do this, but it trips some validation filter that rejects the registration because the prefix is reserved. Presumably this is because we expect the metadata to be added internally to a consul instance rather than coming in from a normal registration. I couldn't figure out an easy way around that but am totally open to suggestions or pointers.

Comment (rboyer) [reply agent/structs/structs.go]: Pushed a commit that refactored how the metadata validation is done so that `consul-wan-federation` is allowed _just_ for service registrations of kind=mesh-gateway.

File: agent/config/builder.go [hunk 22]
```diff
@@ -618,6 +618,7 @@ func (b *Builder) Build() (rt RuntimeConfig, err error) {
 	connectEnabled := b.boolVal(c.Connect.Enabled)
 	connectCAProvider := b.stringVal(c.Connect.CAProvider)
 	connectCAConfig := c.Connect.CAConfig
+	connectMeshGatewayWANFederationEnabled := connectEnabled && b.boolVal(c.Connect.MeshGatewayWANFederationEnabled)
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/config/builder.go]: Changed it to a validation error just to avoid confusion in the case where you omit the `connect.enabled=true` knob.

File: agent/config/flags.go [hunk 23]
```diff
@@ -97,6 +97,8 @@ func AddFlags(fs *flag.FlagSet, f *Flags) {
 	add(&f.Config.RPCProtocol, "protocol", "Sets the protocol version. Defaults to latest.")
 	add(&f.Config.RaftProtocol, "raft-protocol", "Sets the Raft protocol version. Defaults to latest.")
 	add(&f.Config.DNSRecursors, "recursor", "Address of an upstream DNS server. Can be specified multiple times.")
+	add(&f.Config.PrimaryGateways, "primary-gateways", "Address of a mesh gateway in the primary datacenter to use to bootstrap WAN federation at start time with retries enabled. Can be specified multiple times.")
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/config/flags.go]: Yeah that makes sense to change. I almost want to just get rid of `-primary-gateways-interval` as a flag (leaving it just as a config option) while making changes. I doubt it's going to be manipulated enough to warrant a flag, and this way we don't have to figure out the symmetry here.

I originally only added it as a flag to mimic the way the two retry-joins showed up here.

File: agent/consul/federation_state_endpoint.go (hunk 10 above)
Comment (rboyer) [reply agent/consul/federation_state_endpoint.go]: Well if there was a strong reason I probably would've written a test. I commented this validation out and everything important passed so I can't remember why I thought this was important to enforce.

File: agent/consul/gateway_locator.go (hunk 12 above)
Comment (rboyer) [reply agent/consul/gateway_locator.go]: I clarified the comment.

File: agent/consul/gateway_locator.go (hunk 13 above)
Comment (rboyer) [reply agent/consul/gateway_locator.go]: It's got the weird signature so that it can work in a retryJoin.join callback. I'm going to clarify this by confining the weirdness just to where that callback is used instead of spreading it all over.

File: agent/consul/gateway_locator.go (hunk 14 above)
Comment (rboyer) [reply agent/consul/gateway_locator.go]: It was for testing, to avoid having to spin up an entire server just to test the light logic in here.

File: agent/consul/gateway_locator.go (hunk 15 above)
Comment (rboyer) [reply agent/consul/gateway_locator.go]: Definitely a :bug:

The intent with this mechanism hooked into the primary-gateway retry-join operation is to have the gateway retry-join loop abort if we find replicated data for gateways, in the event that the retry-join component keeps looping.

This is a bit important as the retry-join-wan loop is blocked from starting to do its thing until _after_ we know we have gateways configured (to avoid useless error logging of the "you can't get there from here" kind).

I can trivially switch this out to a `sync.Once(func() { close(...) })` instead but then I have to figure out how to prove the mechanism I _wanted_ actually is working because clearly it never did.

Comment (rboyer) [reply agent/consul/gateway_locator.go]: Thinking about this more I realized there's a recovery-type situation that we probably need to have a story around (can happen in a different PR) at some point.

1. user initializes with the `-primary-gateway` args
2. system uses fallback addresses from (1) to do the initial wanjoin
3. after wanjoin completes and the rest of the secondary bootstrapping happens we can replicate fedstates
4. now we can exclusively use the primary's fedstate list of gateways instead of the fallback.

But...what if there was some non-graceful rotation of the primary gateways that didn't give the secondaries time to replicate at least one of the new addresses in time? Then the secondary replicated fedstate information is useless AND since having it trumps the flag/config fallback data it will try to use gateways that don't exist "forever".

Probably need some sort of active process in the GatewayLocator that does some simple light RPC (status check?) to the wildcard address through each FSM and fallback gateway in the primary periodically and heuristically determine if the FSM gateways are still usable, and if not goes back to using the fallback ones until the FSM ones work again.

Comment (rboyer) [reply agent/consul/gateway_locator.go]: https://github.com/hashicorp/consul/issues/7339

File: agent/consul/leader_federation_state_ae.go [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+
+			curr := &structs.FederationState{
+				Datacenter:   s.config.Datacenter,
+				MeshGateways: raw,
+			}
+
+			if idx2 > idx1 {
+				queryMeta.Index = idx2
+			} else {
+				queryMeta.Index = idx1
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/consul/leader_federation_state_ae.go]: So what I was trying to do is a blocking query that would unblock if either the fedstate data or the servicedump data was updated. The fedstate data would update due to the replicator leader process and the servicedump data would update if something about the local mesh gateways updated.

These two sets of things use the different indexes so each of these queries may return completely different index values.

After looking at this again just now I'm still not even sure this does what I kinda wanted, nor do I really know how to actually do an apples-and-oranges kind of blocking query.

[the remark this answers is not in the document]
Comment (rboyer) [reply agent/consul/leader_federation_state_ae.go]: So I wrote a test (`TestLeader_FederationStateAntiEntropy_BlockingQuery`) just for the way I expected this blocking query to function and with this idx1/idx2 conditional code it works a-ok.

If I arbitrarily pick either idx1 or idx2 for the returned index from the query (for use in the next iteration) the test fails due to timeout.

If I instrument the conditional during a normal run of the test I see:
```
 idx1=0 idx2=0
 idx1=10 idx2=0
 idx1=10 idx2=0
 idx1=10 idx2=0
 idx1=10 idx2=0
 idx1=10 idx2=0
 idx1=10 idx2=11
 idx1=10 idx2=11
 idx1=12 idx2=11
 idx1=12 idx2=11
 idx1=12 idx2=13
 idx1=12 idx2=13
 idx1=14 idx2=13
```

File: agent/consul/rpc.go [hunk 25]
```diff
@@ -257,6 +370,63 @@ func (s *Server) handleSnapshotConn(conn net.Conn) {
...
+		if err := transport.IngestPacket(lc, conn.RemoteAddr(), time.Now()); err != nil {
+			return err
+		}
+	}
+}
+
+func readUint32(conn net.Conn, timeout time.Duration) (uint32, error) {
+	// Since requests are framed we can easily just set a deadline on
+	// reading that frame and then disable it for the rest of the body.
+	if err := conn.SetDeadline(time.Now().Add(timeout)); err != nil {
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/consul/rpc.go]: done

File: agent/consul/rpc.go [hunk 26]
```diff
@@ -257,6 +370,63 @@ func (s *Server) handleSnapshotConn(conn net.Conn) {
...
+	return v, nil
+}
+
+type limitedConn struct {
+	net.Conn
+	lr io.Reader
+}
+
+func (c *limitedConn) Read(b []byte) (n int, err error) { return c.lr.Read(b) }
+func (c *limitedConn) Close() error                     { return nil /* ignore */ }
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/consul/rpc.go]: Fixed the override of `Close` with a modification to the ingestpacket method to add `shouldClose bool`.
We do still need to pass a `net.Conn` through to that method so we do still need to have an `io.LimitReader` take over for the `Read` method on an existing `net.Conn` though.

File: agent/consul/snapshot_endpoint.go [hunk 27]
```diff
@@ -189,10 +189,21 @@ RESPOND:
...
+	connPool *pool.ConnPool,
+	dc string,
+	nodeName string,
+	addr net.Addr,
+	useTLS bool,
+	args *structs.SnapshotRequest,
+	in io.Reader,
+	reply *structs.SnapshotResponse,
+) (io.ReadCloser, error) {
+	// TODO(rb): should this work dc-to-dc via a mesh gateway?
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/consul/snapshot_endpoint.go]: this appears to work!
```
node-in-dc1$ consul snapshot save -datacenter=dc2 dc2.data
```

File: agent/consul/wanfed/pool.go [hunk 28]
```diff
@@ -0,0 +1,255 @@
...
+		return nil, nil
+
+	case 1:
+		c := conns[0]
+		delete(p.pool, key)
+		return c, nil
+
+	default:
+		sz := len(conns)
+		remaining, last := conns[0:sz-1], conns[sz-1] // car, cdr
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/consul/wanfed/pool.go]: Also turns out even the comment was wrong (should be "cdr car")

File: agent/consul/leader_federation_state_ae.go [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+			if err != nil {
+				return err
+			}
+
+			curr := &structs.FederationState{
+				Datacenter:   s.config.Datacenter,
+				MeshGateways: raw,
+			}
+
+			if idx2 > idx1 {
```
Comment (rboyer) [inline agent/consul/leader_federation_state_ae.go]: comment about this is just MAX()

File: agent/federation_state_endpoint.go [hunk 30]
```diff
@@ -0,0 +1,88 @@
+package agent
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+func (s *HTTPServer) FederationStateGet(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/federation_state_endpoint.go]: fixed

File: agent/pool/conn.go [hunk 31]
```diff
@@ -28,3 +51,27 @@ const (
 	// requiring coordination. RFC 7983 does the marking and goes into
 	// some details about multiplexing connections and identifying TLS.
 )
+
+const RPCMaxTypeValue = 7
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/pool/conn.go]: updated

File: agent/pool/peek_conn.go [hunk 32]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2017 Google Inc.
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/pool/peek_conn.go]: sure

File: agent/pool/pool.go [hunk 33]
```diff
@@ -267,27 +282,90 @@ type HalfCloser interface {
...
+	nodeName string,
+	addr net.Addr,
+	timeout time.Duration,
+	useTLS bool,
+	actualRPCType RPCType,
+) (net.Conn, HalfCloser, error) {
 	p.once.Do(p.init)
 
-	return DialTimeoutWithRPCType(dc, addr, p.SrcAddr, timeout, useTLS || p.ForceTLS, p.TLSConfigurator.OutgoingRPCWrapper(), RPCTLS)
+	if p.GatewayResolver != nil && p.TLSConfigurator != nil && dc != p.Datacenter {
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/pool/pool.go]: Made the connpool aware of client vs server mode and gated the entire conditional on that as well. Also added some commentary.

File: agent/pool/pool.go [hunk 34]
```diff
@@ -327,27 +405,95 @@ func DialTimeoutWithRPCType(dc string, addr net.Addr, src *net.TCPAddr, timeout
 		}
 	}
 
+	if actualRPCType != RPCTLSInsecure || tlsRPCType != RPCTLSInsecure { // Don't send two bytes, insecure TLS isn't wrapped.
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/pool/pool.go]: fixed

File: agent/structs/federation_state.go (hunk 18 above)
Comment (rboyer) [reply agent/structs/federation_state.go]: done

File: agent/consul/rpc.go (hunk 21 above)
Comment (rboyer) [reply agent/consul/rpc.go]: 4MB it is

**Comment by ojagodzinski** [comment]
fails on golang 1.12

```
github.com/hashicorp/consul/agent/consul/gateway_locator.go:170:7: undefined: errors.Is
```

errors.Is introduced in golang 1.13 https://blog.golang.org/go1.13-errors

**Comment by rboyer** [comment]
Consul [is now](https://github.com/hashicorp/consul/pull/7262) built with Go version `1.13.7` so this is to be expected.

**Comment by ojagodzinski** [comment]
so `go 1.12` in go.mod file - https://github.com/hashicorp/consul/blob/master/go.mod#L3 should be also updated to correctly indicate minimum supported Go version.

**Comment by rboyer** [comment]
That directive is a little bit confusing. It is not which version of the compiler and standard library to use, but rather which language features are available when compiling the module. There is a pretty good write up of this directive here: https://blog.ksub.org/bytes/2019/09/15/go.mods-go-directive/

**Comment by ojagodzinski** [comment]
Yes but error message when compiling with 1.12 will be much more informative than IT is now like in linked post:

```
$ go version
go version go1.12.9 linux/amd64

$ go run .
# version-2
./version.go:6:36: syntax error: unexpected b101010, expecting comma or )
note: module requires Go 1.13
```

**Comment by rboyer** [comment]
After discussing with the team we determined that there wasn't any harm to updating the expected language version for the main `github.com/hashicorp/consul` module and it only improved the errors that could surface: https://github.com/hashicorp/consul/pull/7429

**Comment by ojagodzinski** [comment]
thx
